### PR TITLE
Bytecode tests

### PR DESCRIPTION
**Pull request by Daniil-F**
- removed obsolete commands from `ParsePDF`
- made running bytecode throw exceptions so it can be tested
- added tests for byte code
- bugs fixed

### CODE REVIEW

File: Sources/BCCommon/Commons.swift [hunk 1]
```diff
@@ -10,6 +10,17 @@ public struct BCDashPattern {
   }
 }
 
+public struct BCCubicCurve {
+  public let control1: CGPoint
+  public let control2: CGPoint
+  public let to: CGPoint
+  public init(_ control1: CGPoint, _ control2: CGPoint, _ to: CGPoint) {
```
Comment (zienag) [inline Sources/BCCommon/Commons.swift:17]: Explicit arguments names?

File: Sources/BCRunner/BytecodeReaders.swift [hunk 2]
```diff
@@ -4,78 +4,84 @@ import Foundation
 import BCCommon
 
 protocol BytecodeElement {
-  static func readFrom(_ runner: BytecodeRunner) -> Self
+  static func readFrom(_ runner: BytecodeRunner) throws -> Self
 }
 
+struct EnumerationError<T>: Swift.Error {}
```
Comment (zienag) [inline Sources/BCRunner/BytecodeReaders.swift:10]: It’s good to provide some some additional context, like rawValue. Also, EnumerationError is not very good name, maybe something like InvalidRawValue?

File: Sources/BCRunner/BytecodeReaders.swift [hunk 3]
```diff
@@ -4,78 +4,84 @@ import Foundation
...
+    try runner.readInt()
   }
 }
 
 extension UInt8: BytecodeElement {}
 extension UInt32: BytecodeElement {}
 extension UInt16: BytecodeElement {}
 
+extension BytecodeElement where Self: RawRepresentable,
+  Self.RawValue: FixedWidthInteger {
```
Comment (zienag) [inline Sources/BCRunner/BytecodeReaders.swift:23]: There is no need for explicit Self, afaik. And in function body too.

File: Sources/BCRunner/BytecodeRunner.swift [hunk 4]
```diff
@@ -8,6 +8,13 @@ extension BCRGBAColor {
 }
 
 class BytecodeRunner {
+  enum Error: Swift.Error {
```
Comment (zienag) [inline Sources/BCRunner/BytecodeRunner.swift:11]: Nice!

File: Sources/BCRunner/BytecodeRunner.swift [hunk 5]
```diff
@@ -221,35 +249,46 @@ class BytecodeRunner {
...
-  BytecodeRunner(state, commons).run()
+  try BytecodeRunner(state, commons).run()
+}
+
+@_cdecl("runBytecode") func runBytecode(
+  _ context: CGContext,
+  _ start: UnsafePointer<UInt8>,
+  _ len: Int
+) {
+  try! runBytecodeThrowing(context, start, len)
```
Comment (zienag) [inline Sources/BCRunner/BytecodeRunner.swift:293]: Lets be more forgiving, maybe assert?

File: Sources/libcggen/BCCGGenerator.swift [hunk 6]
```diff
@@ -37,7 +37,7 @@ extension Array: ByteCodable where Element: ByteCodable {
 
 extension CGRect: ByteCodable {
   var byteCode: [UInt8] {
-    [x, y, width, height].flatMap(\.byteCode)
+    origin.byteCode + size.byteCode
```
Comment (zienag) [inline Sources/libcggen/BCCGGenerator.swift:40]: 👍🏻

File: Sources/libcggen/PDFToDrawRouteConverter.swift [hunk 7]
```diff
@@ -139,10 +139,10 @@ extension PDFOperator {
       return .curveTo(p1, p2, p3)
     case let .concatCTM(transform):
       return .concatCTM(transform)
-    case .colorSpaceStroke:
-      return .strokeColorSpace
-    case .colorSpaceNonstroke:
-      return .fillColorSpace
+//    case .colorSpaceStroke:
+//      return .strokeColorSpace
```
Comment (zienag) [inline Sources/libcggen/PDFToDrawRouteConverter.swift:143]: You can return .empty here

File: Tests/RegressionTests/BCTests.swift [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+import XCTest
+
+import BCRunner
+import libcggen
+
+private let defTolerance = 0.002
+private let defScale: CGFloat = 2.0
+private let defSize = CGSize(width: 50, height: 50)
+
+let PDFSamplesPath = getCurentFilePath().appendingPathComponent("pdf_samples")
```
Comment (zienag) [inline Tests/RegressionTests/BCTests.swift:12]: Variable names start with lowercase

File: Tests/RegressionTests/SVGTests.swift [hunk 9]
```diff
@@ -293,7 +293,7 @@ extension WKWebView {
   }
 }
 
-private class WKWebViewSnapshoter {
+class WKWebViewSnapshoter {
```
Comment (zienag) [inline Tests/RegressionTests/SVGTests.swift:296]: Probably move to Support.swift?

**Review by zienag — CHANGES_REQUESTED** [review]

File: Sources/BCCommon/Commons.swift (hunk 1 above)
Comment (Daniil-F) [reply Sources/BCCommon/Commons.swift:17]: rn it is consistent with all other types declared in Commons.swift.

File: Sources/BCRunner/BytecodeRunner.swift (hunk 5 above)
Comment (Daniil-F) [reply Sources/BCRunner/BytecodeRunner.swift:293]: wdym?

File: Tests/RegressionTests/SVGTests.swift (hunk 9 above)
Comment (Daniil-F) [reply Tests/RegressionTests/SVGTests.swift:296]: It is for SVG so SVGTests seems to be appropriate place, just like with `renderPDF()`

File: Sources/BCRunner/BytecodeRunner.swift (hunk 5 above)
Comment (zienag) [reply Sources/BCRunner/BytecodeRunner.swift:293]:
```
try {
  runBytecodeThrowing(context, start, len)
} catch {
  assertionFailure("Bytecode runner failed: \(error)")
}
```

File: Tests/RegressionTests/SVGTests.swift (hunk 9 above)
Comment (zienag) [reply Tests/RegressionTests/SVGTests.swift:296]: `*Tests.swift` files are appropriate for tests and for private helpers that are used in those tests. Code that is shared between different `*Tests.swift` files should be extracted to `Support.swift` or something similar like `PDFTestSupport.swift`. `renderPDF()` must've been `fileprivate` too, my bad.

File: Sources/BCCommon/Commons.swift (hunk 1 above)
Comment (zienag) [reply Sources/BCCommon/Commons.swift:17]: And this has already brought us bad luck. That's a sign to be more explicit.

**Review by zienag — APPROVED** [review]

**Pull request merged**